File: dart/leet_code/3075.dart
/// Copyright ©2012-2024 Jerod Gawne <https://github.com/jerodg/>
///
/// This program is free software: you can redistribute it and/or modify it under the terms of the
/// Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
/// or (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
/// even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
/// for more details.
///
/// The above copyright notice and this permission notice shall be included in all copies or
/// substantial portions of the Software. You should have received a copy of the SSPL along with this
/// program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.

/// A `Solution` class that provides a method to calculate the maximum happiness sum.
class Solution {
  /// Calculates the maximum happiness sum from a list of happiness values.
  ///
  /// The function sorts the happiness list in ascending order and then iteratively adds the maximum
  /// happiness value to the sum, decreasing the value by the index each time. If the calculated value
  /// is negative, it adds zero to the sum instead. This process is repeated `k` times.
  ///
  /// @param happiness A list of integers representing happiness values.
  /// @param k The number of times to add the maximum happiness value to the sum.
  /// @return The maximum happiness sum as an integer.
  int maximumHappinessSum(List<int> happiness, int k) {
    happiness.sort();
    int sum = 0;
    int index = 0;

    while (k > 0) {
      // Calculate the value to add to the sum. If the value is negative, add zero instead.
      sum += (happiness[happiness.length - index - 1] - index).isNegative
          ? 0
          : (happiness[happiness.length - index - 1] - index);
      k--;
      index++;
    }
    return sum;
  }
}

File: ruby/leet_code/3075.rb
# Copyright © 2010-2024 JerodG <https://github.com/jerodg/> 2010-2024.
#
# This program is free software: you can redistribute it and/or modify it under the terms of the
# Server Side Public License (SSPL) as published by MongoDB, Inc., either version 1 of the License,
# or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
# even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the SSPL
# for more details.
#
# The above copyright notice and this permission notice shall be included in all copies or
# substantial portions of the Software. You should have received a copy of the SSPL along with this
# program. If not, see <https://www.mongodb.com/licensing/server-side-public-license>.

# frozen_string_literal: true

# This file contains a method `maximum_happiness_sum` which calculates the maximum happiness sum
# given an array of happiness values and a number `k`. The method sorts the happiness array in
# ascending order, then iteratively pops the highest happiness value, subtracts the current
# iteration index from it, and adds it to the total sum.
# The iteration stops if the calculated happiness value becomes negative.

# @param [Array<Integer>] happiness An array of integers representing happiness values.
# @param [Integer] k The number of times to subtract the iteration index from the highest happiness value.
# @return [Integer] The maximum happiness sum.
def maximum_happiness_sum(happiness, k)
  # Sort the happiness array in ascending order
  happiness.sort!

  # Initialize the total sum to 0
  total = 0

  # Iterate `k` times
  k.times do |i|
    # Pop the highest happiness value
    h = happiness.pop

    # Subtract the current iteration index from the popped happiness value
    l = h - i

    # Break the loop if the calculated happiness value is negative
    break if l.negative?

    # Add the calculated happiness value to the total sum
    total += l
  end

  # Return the total sum
  total
end
